### PR TITLE
Avoid deadlocks in TypeDescriptor

### LINKED ISSUE #97179: Potential deadlock in TypeDescriptor
Introduced by the recent PR https://github.com/dotnet/runtime/pull/96846.

There are two sync objects being used (`s_defaultProviderSyncObject`, `s_providerTable`) and a deadlock is possible since the ordering of these 2 locks can vary and both locks rerequested.

The repro is to call `TypeDescriptor.GetProvider(Type)` and `TypeDescriptor.AddProvider(..., Type) on different threads.

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/97179

The new test deadlocked about half of the time before the change here.

### CODE REVIEW

**Review by ericstj — APPROVED** [review]
Nice catch, fix looks reasonable.

**Review by ericstj — COMMENTED** [review]
Actually look like this is failing validation tests added with previous PR, seeing if I can spot what might be causing it.

File: src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs [hunk 1]
```diff
@@ -1289,6 +1290,57 @@ public void ConcurrentGetProperties_ReturnsExpected()
...
+        {
+            // Use a timeout value lower than RemoteExecutor in order to get a nice Fail message.
+            const int Timeout = 50000;
+
+            RemoteInvokeOptions options = new()
+            {
+                TimeOut = 60000
+            };
+
+            ConcurrentAddProvider();
```
Comment (ericstj) [inline src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs:1305]: You call this in the test process, which ends up registering `EmptyPropertiesTypeProvider` for `typeof(MyClass)`, this breaks the remaining tests (GetConverter_ByMultithread_ReturnsExpected) which run in process and expect the provider registered with `TypeDescriptionProviderAttribute`.

What made you call `ConcurrentAddProvider` in proc?  Seems like the rest of the test is happening out of proc.

Comment (ericstj) [inline src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs:1305]:
```suggestion
```

You call this in the main test process, which ends up registering `EmptyPropertiesTypeProvider` for `typeof(MyClass)`, this breaks the remaining tests (`GetConverter_ByMultithread_ReturnsExpected`) which run in process and expect the provider registered with `TypeDescriptionProviderAttribute`.

What made you call `ConcurrentAddProvider` in proc?  Seems like the rest of the test is happening out of proc.

**Review by stephentoub — COMMENTED** [review]
Fix LGTM other than Eric's comments on the tests.

File: src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs]: Oops - I added that temporarily for debugging purposes due to RemoteExecutor below that. Thanks

**Comment by steveharter** [comment]
CI failure unrelated: https://github.com/dotnet/runtime/issues/97241

**Comment by ericstj** [comment]
LGTM.  Thanks for catching this.

**Pull request merged**